torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1086,6 +1098,38 @@ def _assert_state(self, state: Union[TrainingState_, List[TrainingState_]]) -> N
                 traceback.print_stack()
             raise ValueError(msg)
 
+    @contextmanager
+    def no_sync(self) -> Generator:
+        """
+        A context manager to disable gradient synchronizations across FSDP
+        instances. Within this context, gradients will be accumulated on module
+        variables, which will later be synchronized in the first
+        forward-backward pass after exiting the context.
+
+        .. note:: This likely results in higher memory usage because FSDP will
+            accumulate the full model gradients (instead of gradient shards)
+            until the eventual sync.
+
+        .. note:: Gradient accumulation can be done without this context,
+            avoiding the extra GPU memory overhead, but with the extra
+            networking overhead.
+        """
+        self._lazy_init()
+        assert self._is_root, \
+            "`no_sync()` on inner FSDP instances is not supported"
+        self._assert_state(TrainingState_.IDLE)
+        old_flags = []
+        for m in self.modules():
+            if isinstance(m, FullyShardedDataParallel):
+                old_flags.append((m, m._require_backward_grad_sync))
```

Comment:
This is minor, but in my understanding, Python generator ordering is not guaranteed. In practice, `torch.nn.Module.modules()` probably does always return the same ordering, but this might be relying on non-specified behavior.

Is there any downside to maintaining an extra reference to each module? The modules should not be being deleted anyway (so these references should not be the critical ones preventing garbage collection)?